File: app/src/main/java/com/velaphi/untamed/features/animalDetails/models/Converters.java
package com.velaphi.untamed.features.animalDetails.models;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.List;

public class Converters {

    @TypeConverter
    public static List<String> restoreList(String listOfString) {
        return new Gson().fromJson(listOfString, new TypeToken<List<String>>() {
        }.getType());
    }

    @TypeConverter
    public static String saveList(List<String> listOfString) {
        return new Gson().toJson(listOfString);
    }


    @TypeConverter
    public static String fromBasicInfoList(List<Info> infoList) {
        return new Gson().toJson(infoList);
    }

    @TypeConverter
    public static List<Info> toBasicInfoList(String basicInfoJson) {
        return new Gson().fromJson(basicInfoJson, new TypeToken<List<Info>>() {
        }.getType());
    }

    @TypeConverter
    public static String fromPredatorList(List<Predator> predatorList) {
        return new Gson().toJson(predatorList);
    }

    @TypeConverter
    public static List<Predator> toPredatorList(String predatorListJson) {
        return new Gson().fromJson(predatorListJson, new TypeToken<List<Predator>>() {
        }.getType());
    }

    @TypeConverter
    public static List<Prey> toPreyList(String preyListJson) {
        return new Gson().fromJson(preyListJson, new TypeToken<List<Prey>>() {
        }.getType());
    }

    @TypeConverter
    public static String fromPreyList(List<Prey> preyList) {
        return new Gson().toJson(preyList);
    }


    @TypeConverter
    public static String fromFactList(List<Fact> factList) {
        return new Gson().toJson(factList);
    }

    @TypeConverter
    public static List<Fact> toFactList(String factJson) {
        return new Gson().fromJson(factJson, new TypeToken<List<Fact>>() {
        }.getType());
    }

    @TypeConverter
    public static String fromChallengeList(List<Challenge> challengeList) {
        return new Gson().toJson(challengeList);
    }

    @TypeConverter
    public static List<Challenge> toChallengetList(String challengesJson) {
        return new Gson().fromJson(challengesJson, new TypeToken<List<Challenge>>() {
        }.getType());
    }


    @TypeConverter
    public static String fromVideoList(List<Video> videoList) {
        return new Gson().toJson(videoList);
    }

    @TypeConverter
    public static ArrayList<Video> toVideoList(String videosJson) {
        return new Gson().fromJson(videosJson, new TypeToken<List<Video>>() {
        }.getType());
    }


    @TypeConverter
    public static Habitat restoreHabit(String habitatString) {
        return new Gson().fromJson(habitatString, Habitat.class);
    }

    @TypeConverter
    public static String saveHabitatToJson(Habitat habitat) {
        return new Gson().toJson(habitat);
    }

    @TypeConverter
    public static Description restoreDescription(String descriptionString) {
        return new Gson().fromJson(descriptionString, Description.class);
    }

    @TypeConverter
    public static String saveDescriptionToJson(Description description) {
        return new Gson().toJson(description);
    }
}

File: app/src/main/java/com/velaphi/untamed/features/animalDetails/models/Converters.kt
package com.velaphi.untamed.features.animalDetails.models

import androidx.room.TypeConverter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.*

object Converters {
    @JvmStatic
    @TypeConverter
    fun restoreList(listOfString: String?): List<String> {
        return Gson().fromJson(listOfString, object : TypeToken<List<String?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun saveList(listOfString: List<String?>?): String {
        return Gson().toJson(listOfString)
    }

    @JvmStatic
    @TypeConverter
    fun fromBasicInfoList(infoList: List<Info?>?): String {
        return Gson().toJson(infoList)
    }

    @JvmStatic
    @TypeConverter
    fun toBasicInfoList(basicInfoJson: String?): List<Info> {
        return Gson().fromJson(basicInfoJson, object : TypeToken<List<Info?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun fromBasicCountryList(countryList: List<Country?>?): String {
        return Gson().toJson(countryList)
    }

    @JvmStatic
    @TypeConverter
    fun toCountryList(countryListJson: String?): List<Country> {
        return Gson().fromJson(countryListJson, object : TypeToken<List<Country?>?>() {}.type)
    }
    @JvmStatic
    @TypeConverter
    fun fromPredatorList(predatorList: List<Predator?>?): String {
        return Gson().toJson(predatorList)
    }

    @JvmStatic
    @TypeConverter
    fun toPredatorList(predatorListJson: String?): List<Predator> {
        return Gson().fromJson(predatorListJson, object : TypeToken<List<Predator?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun toPreyList(preyListJson: String?): List<Prey> {
        return Gson().fromJson(preyListJson, object : TypeToken<List<Prey?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun fromPreyList(preyList: List<Prey?>?): String {
        return Gson().toJson(preyList)
    }

    @JvmStatic
    @TypeConverter
    fun fromFactList(factList: List<Fact?>?): String {
        return Gson().toJson(factList)
    }

    @JvmStatic
    @TypeConverter
    fun toFactList(factJson: String?): List<Fact> {
        return Gson().fromJson(factJson, object : TypeToken<List<Fact?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun fromChallengeList(challengeList: List<Challenge?>?): String {
        return Gson().toJson(challengeList)
    }

    @JvmStatic
    @TypeConverter
    fun toChallengeList(challengesJson: String?): List<Challenge> {
        return Gson().fromJson(challengesJson, object : TypeToken<List<Challenge?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun fromVideoList(videoList: List<Video?>?): String {
        return Gson().toJson(videoList)
    }

    @JvmStatic
    @TypeConverter
    fun toVideoList(videosJson: String?): ArrayList<Video> {
        return Gson().fromJson(videosJson, object : TypeToken<List<Video?>?>() {}.type)
    }

    @JvmStatic
    @TypeConverter
    fun restoreHabit(habitatString: String?): Habitat {
        return Gson().fromJson(habitatString, Habitat::class.java)
    }

    @JvmStatic
    @TypeConverter
    fun saveHabitatToJson(habitat: Habitat?): String {
        return Gson().toJson(habitat)
    }

    @JvmStatic
    @TypeConverter
    fun restoreDescription(descriptionString: String?): Description {
        return Gson().fromJson(descriptionString, Description::class.java)
    }

    @JvmStatic
    @TypeConverter
    fun saveDescriptionToJson(description: Description?): String {
        return Gson().toJson(description)
    }
}